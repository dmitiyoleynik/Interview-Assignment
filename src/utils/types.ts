export type Ship = {
  id: number;
  built: number;
  name: string;
  lengthMeters: number;
  beamMeters: number;
  maxTEU: number;
  owner: string;
  grossTonnage: number | string;
};
