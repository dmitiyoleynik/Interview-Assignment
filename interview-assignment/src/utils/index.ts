import { useMediaQuery, useTheme } from '@mui/material';

export const getCountryOfOrigin = (shipOwner: string) => {
  const regex = /\(([^)]+)\)/;
  const matches = regex.exec(shipOwner);

  return matches && matches[1];
};

type size = 'xxs' | 'xs' | 'sm' | 'md' | 'lg';

type sizes<T> = {
  xxs: T;
  xs: T;
  sm: T;
  md: T;
  lg: T;
};

export const useFlexibleSizes = <T>(sizes: sizes<T>) => {
  const theme = useTheme();

  const breakpointsMap: { useMediaQueryResult: boolean; type: size }[] = [
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up('lg')),
      type: 'lg',
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up('md')),
      type: 'md',
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up('sm')),
      type: 'sm',
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up(280)),
      type: 'xs',
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.down(280)),
      type: 'xxs',
    },
  ];
  console.log({ breakpointsMap });

  const matchingBreakpoint = breakpointsMap.find((s) => s.useMediaQueryResult);

  return matchingBreakpoint ? sizes[matchingBreakpoint.type] : sizes.xxs;
};
