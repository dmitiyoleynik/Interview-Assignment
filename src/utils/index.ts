import { useMediaQuery, useTheme } from '@mui/material';

export const getCountryOfOrigin = (shipOwner: string) => {
  const regex = /\(([^)]+)\)/;
  const matches = regex.exec(shipOwner);
  const matchIndex = 1;

  return matches?.[matchIndex];
};

enum Size {
  'xxs' = 'xxs',
  'xs' = 'xs',
  'sm' = 'sm',
  'md' = 'md',
  'lg' = 'lg',
}

type Sizes<T> = {
  [key in Size]: T;
};

export const useFlexibleSizes = <T>(sizes: Sizes<T>) => {
  const theme = useTheme();

  const breakpointsMap: { useMediaQueryResult: boolean; type: Size }[] = [
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up('lg')),
      type: Size.lg,
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up('md')),
      type: Size.md,
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up('sm')),
      type: Size.sm,
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.up(280)),
      type: Size.xs,
    },
    {
      useMediaQueryResult: useMediaQuery(theme.breakpoints.down(280)),
      type: Size.xxs,
    },
  ];
  console.log({ breakpointsMap });

  const matchingBreakpoint = breakpointsMap.find((s) => s.useMediaQueryResult);

  return matchingBreakpoint ? sizes[matchingBreakpoint.type] : sizes.xxs;
};
