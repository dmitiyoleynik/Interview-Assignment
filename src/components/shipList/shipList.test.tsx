import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import '@testing-library/jest-dom';

import { Ship } from 'src/utils/types';
import ShipList from 'src/components/shipList/shipList';

test('ShipList should match snapshot with zero ship', () => {
  const component = renderer.create(
    <MemoryRouter>
      <ShipList ships={[]} />
    </MemoryRouter>,
  );
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test('ShipList should match snapshot with one ship', () => {
  const ships: Ship[] = [
    {
      id: 2,
      built: 2017,
      name: 'OOCL Germany',
      lengthMeters: 399.87,
      beamMeters: 58.8,
      maxTEU: 21413,
      owner: 'OOCL (Hong Kong)',
      grossTonnage: 210890,
    },
  ];
  const component = renderer.create(
    <MemoryRouter>
      <ShipList ships={ships} />
    </MemoryRouter>,
  );
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test('ShipList should match snapshot with many ships', () => {
  const ships: Ship[] = [
    {
      id: 2,
      built: 2017,
      name: 'OOCL Germany',
      lengthMeters: 399.87,
      beamMeters: 58.8,
      maxTEU: 21413,
      owner: 'OOCL (Hong Kong)',
      grossTonnage: 210890,
    },
    {
      id: 3,
      built: 2017,
      name: 'OOCL Japan',
      lengthMeters: 399.87,
      beamMeters: 58.8,
      maxTEU: 21413,
      owner: 'OOCL (Hong Kong)',
      grossTonnage: 210890,
    },
  ];
  const component = renderer.create(
    <MemoryRouter>
      <ShipList ships={ships} />
    </MemoryRouter>,
  );
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
