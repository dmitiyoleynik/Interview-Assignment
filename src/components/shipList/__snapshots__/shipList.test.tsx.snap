// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` MyComponent should match snapshot 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1sdcbww-MuiGrid-root"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
  >
    <div
      className="MuiStack-root css-1z0rh7t-MuiStack-root"
    >
      <a
        className="makeStyles-link-2"
        href="/ships/2"
        onClick={[Function]}
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root makeStyles-card-1 css-uzaqbr-MuiPaper-root-MuiCard-root"
        >
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <h6
              className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              OOCL Germany
            </h6>
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Country: 
              Hong Kong
            </p>
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              TEU: 
              21413
            </p>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[` ShipList should match snapshot 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1sdcbww-MuiGrid-root"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
  >
    <div
      className="MuiStack-root css-1z0rh7t-MuiStack-root"
    >
      <a
        className="makeStyles-link-2"
        href="/ships/2"
        onClick={[Function]}
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root makeStyles-card-1 css-uzaqbr-MuiPaper-root-MuiCard-root"
        >
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <h6
              className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              OOCL Germany
            </h6>
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Country: 
              Hong Kong
            </p>
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              TEU: 
              21413
            </p>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`ShipList should match snapshot 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1sdcbww-MuiGrid-root"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
  >
    <div
      className="MuiStack-root css-1z0rh7t-MuiStack-root"
    >
      <a
        className="makeStyles-link-2"
        href="/ships/2"
        onClick={[Function]}
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root makeStyles-card-1 css-uzaqbr-MuiPaper-root-MuiCard-root"
        >
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <h6
              className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              OOCL Germany
            </h6>
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Country: 
              Hong Kong
            </p>
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              TEU: 
              21413
            </p>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`ShipList should match snapshot with many ships 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1sdcbww-MuiGrid-root"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
  >
    <div
      className="MuiStack-root css-1z0rh7t-MuiStack-root"
    >
      <a
        className="makeStyles-link-2"
        href="/ships/2"
        onClick={[Function]}
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root makeStyles-card-1 css-uzaqbr-MuiPaper-root-MuiCard-root"
        >
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <h6
              className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              OOCL Germany
            </h6>
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Country: 
              Hong Kong
            </p>
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              TEU: 
              21413
            </p>
          </div>
        </div>
      </a>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
  >
    <div
      className="MuiStack-root css-1z0rh7t-MuiStack-root"
    >
      <a
        className="makeStyles-link-2"
        href="/ships/3"
        onClick={[Function]}
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root makeStyles-card-1 css-uzaqbr-MuiPaper-root-MuiCard-root"
        >
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <h6
              className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              OOCL Japan
            </h6>
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Country: 
              Hong Kong
            </p>
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              TEU: 
              21413
            </p>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`ShipList should match snapshot with one ship 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1sdcbww-MuiGrid-root"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-14j1sok-MuiGrid-root"
  >
    <div
      className="MuiStack-root css-1z0rh7t-MuiStack-root"
    >
      <a
        className="makeStyles-link-2"
        href="/ships/2"
        onClick={[Function]}
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root makeStyles-card-1 css-uzaqbr-MuiPaper-root-MuiCard-root"
        >
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <h6
              className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              OOCL Germany
            </h6>
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Country: 
              Hong Kong
            </p>
            <p
              className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              TEU: 
              21413
            </p>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`ShipList should match snapshot with zero ship 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1sdcbww-MuiGrid-root"
/>
`;
